### PR TITLE
[geometry_msgs] Workaround for #430

### PR DESCRIPTION
**Pull request by gleichdick**
Workaround for #430. Snippet now compiles (with convert.h included last)

```c++
#include <tf2/LinearMath/Quaternion.h>
#include <Eigen/Geometry>
#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
#include <tf2_eigen/tf2_eigen.h>
#include <tf2/convert.h>
#include <iostream>

int main() {
  Eigen::Quaterniond eq;

  const tf2::Quaternion tq(0,1,0,1);
  tf2::convert(tq, eq);
  std::cout << eq.toRotationMatrix();
}
```

Note that it still depends on the include order, if convert.h is included first it does not work (compiler picks forward declaration of templated `fromMsg` which is not resolved later on).

### CODE REVIEW

**Comment by v4hn** [comment]
C++ includes should be independent of the include order.
Is it difficult to resolve the problem altogether?

**Comment by gleichdick** [comment]
**TL;DR** Please a look at [my branch](https://github.com/gleichdick/geometry2/tree/templated_fromMsg). `toMsg` is redefined (so, client code which uses `toMsg` breaks) but `tf2::convert` should be fine. Include order does not matter anymore.


I converted all `toMsg` overloads to template specialisations with two parameters (because the return value does not count for template parameter deduction). All overloads (without `template<>`) are gone:
```c++
template<class A, class B> B& toMsg(const A&, B&);
```

For converting two non-Message types (lets say `tf2::Vector3` and `Eigen::Vector3d`), I'm writing a python script which generates type mapping (both have conversions to/from `geometry_msgs::Point` in common), like
```c++
template<>
struct commonMsgType<tf2::Vector3, Eigen::Vector3> {
    using type = geometry_msgs::Point;
};
```
This struct is then read by `tf2::convert` with some SFINAE Magic which then selects the correct `fromMsg` / `toMsg` methods.

Please note that this branch is more or less a proof of concept, so documentation and tests are missing. I'm happy to contine, but please let me know whether this solution fits your needs.

**Comment by gleichdick** [comment]
replaced by #491
